bigframes/core/rewrite.py

```diff
@@ -0,0 +1,305 @@
+        if isinstance(node, nodes.ProjectionNode):
+            return cls.from_node(node.child).project(node.assignments)
+        elif isinstance(node, nodes.FilterNode):
+            return cls.from_node(node.child).filter(node.predicate)
+        elif isinstance(node, nodes.ReversedNode):
+            return cls.from_node(node.child).reverse()
+        elif isinstance(node, nodes.OrderByNode):
+            return cls.from_node(node.child).order_with(node.by)
+        else:
+            selection = tuple(
+                (scalar_exprs.UnboundVariableExpression(id), id)
+                for id in get_node_column_ids(node)
+            )
+            return cls(node, selection, None, ())
+
+    def project(
+        self, projection: Tuple[Tuple[scalar_exprs.Expression, str], ...]
+    ) -> NormalizedSelect:
+        lookup = {id: expr for expr, id in self.columns}
+        new_columns = tuple(
+            (expr.bind_variables(lookup, bind_all=True), id) for expr, id in projection
+        )
+        return NormalizedSelect(self.root, new_columns, self.predicate, self.ordering)
+
+    def filter(self, predicate: scalar_exprs.Expression) -> NormalizedSelect:
+        lookup = {id: expr for expr, id in self.columns}
+        if self.predicate is None:
+            new_predicate = predicate.bind_variables(lookup, bind_all=True)
+        else:
+            new_predicate = ops.and_op.as_expr(
+                self.predicate, predicate.bind_variables(lookup, bind_all=True)
+            )
+        return NormalizedSelect(self.root, self.columns, new_predicate, self.ordering)
+
+    def reverse(self) -> NormalizedSelect:
+        new_ordering = tuple(expr.with_reverse() for expr in self.ordering)
+        return NormalizedSelect(self.root, self.columns, self.predicate, new_ordering)
+
+    def order_with(self, by: Tuple[order.OrderingColumnReference, ...]):
+        exprs_by_id = {id: expr for expr, id in self.columns}
+        as_order_exprs = [
+            order.OrderingExpression(
+                exprs_by_id[ref.column_id], ref.direction, ref.na_last
+            )
+            for ref in by
+        ]
+        new_ordering = (*as_order_exprs, *self.ordering)
+        return NormalizedSelect(self.root, self.columns, self.predicate, new_ordering)
+
+    def maybe_join(
+        self, right: NormalizedSelect, join_def: join_defs.JoinDefinition
+    ) -> Optional[NormalizedSelect]:
+        if join_def.type == "cross":
+            # Cannot convert cross join to projection
+            return None
+
+        r_exprs_by_id = {id: expr for expr, id in right.columns}
+        l_exprs_by_id = {id: expr for expr, id in self.columns}
+        l_join_exprs = [l_exprs_by_id[cond.left_id] for cond in join_def.conditions]
+        r_join_exprs = [r_exprs_by_id[cond.right_id] for cond in join_def.conditions]
+
+        if (self.root == right.root) and all(
+            l_expr == r_expr for l_expr, r_expr in zip(l_join_exprs, r_join_exprs)
+        ):
```

Comment:
Seems we could reduce some nesting if we did the inverse of this and exit early. go/tott/651

Reply:
done